db/db_merge_operator_test.cc

```diff
@@ -202,6 +202,116 @@ TEST_F(DBMergeOperatorTest, MergeErrorOnIteration) {
   VerifyDBInternal({{"k1", "v1"}, {"k2", "corrupted"}, {"k2", "v2"}});
 }
 
+#ifndef ROCKSDB_LITE
+
+TEST_F(DBMergeOperatorTest, MergeOperatorFailsWithMustMerge) {
+  const int kNumOperands = 3;
+  Options options;
+  options.merge_operator.reset(new TestPutOperator());
+  options.env = env_;
+  Reopen(options);
+
+  auto check_query = [&](Slice key) {
+    {
+      std::string value;
+      ASSERT_TRUE(db_->Get(ReadOptions(), key, &value).IsCorruption());
+    }
+
+    {
+      std::unique_ptr<Iterator> iter;
+      iter.reset(db_->NewIterator(ReadOptions()));
+      iter->SeekToFirst();
+      ASSERT_TRUE(iter->status().IsCorruption());
+    }
+
+    std::vector<PinnableSlice> values(kNumOperands);
+    GetMergeOperandsOptions merge_operands_info;
+    merge_operands_info.expected_max_number_of_operands = kNumOperands;
+    int num_operands_found = 0;
+    ASSERT_OK(db_->GetMergeOperands(ReadOptions(), db_->DefaultColumnFamily(),
+                                    key, values.data(), &merge_operands_info,
+                                    &num_operands_found));
+    ASSERT_EQ(kNumOperands, num_operands_found);
+    for (int i = 0; i < num_operands_found; ++i) {
+      if (i == 0) {
+        ASSERT_EQ(values[i], "corrupted_must_merge");
+      } else {
+        ASSERT_EQ(values[i], "ok");
+      }
+    }
+  };
+
+  // Case 1: Mixed good/bad `Merge()`s
+  for (int i = 0; i < kNumOperands; ++i) {
+    if (i == 0) {
+      ASSERT_OK(Merge("k1", "corrupted_must_merge"));
+    } else {
+      ASSERT_OK(Merge("k1", "ok"));
+    }
+  }
```

Comment:
It might be better also cover the case where there is a "k2" inserted too. There is a risk that in compaction the internal keys aren't advanced correctly and k2 is like missing.

Reply:
Done